[CIR] Upstream SelectOp and ShiftOp

mmha opened the pull request:
Since SelectOp will only generated by a future pass that transforms a TernaryOp this only includes the lowering bits.

This patch also improves the testing of the existing binary operators.

mmha commented:
cc @andykaylor @erichkeane @dkolsen-pgi

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -759,6 +762,46 @@ LogicalResult cir::BinOp::verify() {
+//===----------------------------------------------------------------------===//
+
+OpFoldResult cir::SelectOp::fold(FoldAdaptor adaptor) {
+  mlir::Attribute condition = adaptor.getCondition();
```
Please add tests for the folders.

mmha replied:
I removed the folding from this PR and will readd it along with the rest of the `cir-simplify` pass after TernaryOp lands.

bcardosolopes commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -1117,6 +1117,91 @@ mlir::LogicalResult CIRToLLVMBinOpLowering::matchAndRewrite(
+mlir::LogicalResult CIRToLLVMSelectOpLowering::matchAndRewrite(
+    cir::SelectOp op, OpAdaptor adaptor,
+    mlir::ConversionPatternRewriter &rewriter) const {
+  auto getConstantBool = [](mlir::Value value) -> std::optional<bool> {
```
Let's rely on MLIR's usage of value semantics here.
This lambda should just return an `cir::BoolAttr`, so you can write:
```
cir::BoolAttr trueValue = getConstantBool(op.getTrueValue());
...
if (trueValue && !trueValue.getValue()) {
}
``` 
The lambda should return `{}`/`nullptr` in the case attribute not found.

bcardosolopes reviewed (changes requested):
Looks reasonable to add vector support later for shifts. Because `SelectOp` is independent of `ShiftOp`, one possibility would have been to split it in two PRs to land them faster, but wouldn't bother changing for this round.

andykaylor commented on clang/test/CIR/CodeGen/binop.cpp:
```diff
@@ -12,22 +16,307 @@ void b0(int a, int b) {
+
+// OGCG:         ret void
+
+void signed_shift(int a, int b) {
```
Can you add cases where the shift value needs `zext` and `sext`?

erichkeane reviewed (commented):
Andy/Bruno have a good hold on this, so I'll let them review it.

mmha committed:
[CIR] Upstream SelectOp and ShiftOp

Since SelectOp will only generated by a future pass that transforms a TernaryOp this only includes the lowering bits.

This patchs also improves the testing of the existing binary operators.

mmha committed:
Address review feedback

- add createShift functions to CIRBuilder
- Rephrase ShiftOpn description comment
- Remove folding for now
- Always zero-extend instead of potentially sign-extend
- Tests with diffferently sized integral types

mmha commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -1294,6 +1296,90 @@ mlir::LogicalResult CIRToLLVMCmpOpLowering::matchAndRewrite(
+  // be already be enforced at CIRGen.
+  if (cirAmtTy)
+    amt = getLLVMIntCast(rewriter, amt, mlir::cast<mlir::IntegerType>(llvmTy),
+                         true, cirAmtTy.getWidth(), cirValTy.getWidth());
```
I just changed this to always do zero extension (instead of potentially sign extending if the amount type is signed). This follows the original clang CG behavior and is probably slightly more efficient on some architectures.

mmha replied:
There you go: https://github.com/llvm/clangir/pull/1568

bcardosolopes replied:
Awesome, merged!
